From 70a23a757b32ebdde8b7c22ebed46667381c5125 Mon Sep 17 00:00:00 2001
From: Ziyang Huang <hzyitc@outlook.com>
Date: Sun, 8 Sep 2024 16:52:04 +0800
Subject: [PATCH 1/1] wifi: ath11k: introduct CONFIG_ATH11K_SMALLBUFFERS

Reduce buffer size so we can run on small RAM device.

Signed-off-by: Ziyang Huang <hzyitc@outlook.com>
---
 drivers/net/wireless/ath/ath11k/Kconfig |  6 ++++++
 drivers/net/wireless/ath/ath11k/dp.h    | 10 +++++-----
 local-symbols                           |  1 +
 3 files changed, 12 insertions(+), 5 deletions(-)

--- a/drivers/net/wireless/ath/ath11k/Kconfig
+++ b/drivers/net/wireless/ath/ath11k/Kconfig
@@ -12,6 +12,12 @@ config ATH11K
 
 	  If you choose to build a module, it'll be called ath11k.
 
+config ATH11K_SMALLBUFFERS
+	bool "Use small buffers"
+	depends on ATH11K
+	help
+	  Select this for low-RAM devices.
+
 config ATH11K_AHB
 	tristate "Atheros ath11k AHB support"
 	depends on m
--- a/drivers/net/wireless/ath/ath11k/dp.h
+++ b/drivers/net/wireless/ath/ath11k/dp.h
@@ -205,7 +205,7 @@ struct ath11k_pdev_dp {
 #define DP_WBM_RELEASE_RING_SIZE	64
 #define DP_TCL_DATA_RING_SIZE		512
 #define DP_TCL_DATA_RING_SIZE_WCN6750	2048
-#define DP_TX_COMP_RING_SIZE		32768
+#define DP_TX_COMP_RING_SIZE		(!IS_ENABLED(CPTCFG_ATH11K_SMALLBUFFERS) ? 32768 : 2048)
 #define DP_TX_IDR_SIZE			DP_TX_COMP_RING_SIZE
 #define DP_TCL_CMD_RING_SIZE		32
 #define DP_TCL_STATUS_RING_SIZE		32
@@ -216,12 +216,12 @@ struct ath11k_pdev_dp {
 #define DP_REO_EXCEPTION_RING_SIZE	128
 #define DP_REO_CMD_RING_SIZE		256
 #define DP_REO_STATUS_RING_SIZE		2048
-#define DP_RXDMA_BUF_RING_SIZE		4096
+#define DP_RXDMA_BUF_RING_SIZE		(!IS_ENABLED(CPTCFG_ATH11K_SMALLBUFFERS) ? 4096 : 1024)
 #define DP_RXDMA_REFILL_RING_SIZE	2048
 #define DP_RXDMA_ERR_DST_RING_SIZE	1024
-#define DP_RXDMA_MON_STATUS_RING_SIZE	1024
-#define DP_RXDMA_MONITOR_BUF_RING_SIZE	4096
-#define DP_RXDMA_MONITOR_DST_RING_SIZE	2048
+#define DP_RXDMA_MON_STATUS_RING_SIZE	(!IS_ENABLED(CPTCFG_ATH11K_SMALLBUFFERS) ? 1024 : 512)
+#define DP_RXDMA_MONITOR_BUF_RING_SIZE	(!IS_ENABLED(CPTCFG_ATH11K_SMALLBUFFERS) ? 4096 : 128)
+#define DP_RXDMA_MONITOR_DST_RING_SIZE	(!IS_ENABLED(CPTCFG_ATH11K_SMALLBUFFERS) ? 2048 : 128)
 #define DP_RXDMA_MONITOR_DESC_RING_SIZE	4096
 
 #define DP_RX_RELEASE_RING_NUM	3
--- a/local-symbols
+++ b/local-symbols
@@ -160,6 +160,7 @@ ATH10K_DFS_CERTIFIED=
 WCN36XX=
 WCN36XX_DEBUGFS=
 ATH11K=
+ATH11K_SMALLBUFFERS=
 ATH11K_AHB=
 ATH11K_PCI=
 ATH11K_DEBUG=
